File: tests/test_parse/data/sample.lua
-- Regular function
function add(a, b)
    return a + b
end

-- Local function
local function subtract(a, b)
    return a - b
end

-- Anonymous function assigned to variable
local multiply = function(a, b)
    return a * b
end

-- Function with multiple returns
function divmod(a, b)
    return math.floor(a/b), a % b
end

-- Function with variable arguments
function sum(...)
    local result = 0
    for _, v in ipairs({...}) do
        result = result + v
    end
    return result
end

-- Table with methods
local Calculator = {
    value = 0,
    
    -- Method using colon syntax
    add = function(self, x)
        self.value = self.value + x
    end,
    
    -- Alternative method syntax
    subtract = function(self, x)
        self.value = self.value - x
    end
}

-- Method using colon syntax (syntactic sugar)
function Calculator:multiply(x)
    self.value = self.value * x
end

-- Class-like structure with constructor
local function createPerson(name)
    local person = {
        name = name,
        
        -- Method
        greet = function(self)
            return "Hello, I'm " .. self.name
        end
    }
    
    -- Return the instance
    return person
end

-- Closure example
function counter()
    local count = 0
    return function()
        count = count + 1
        return count
    end
end

-- Function that returns multiple functions
function makeOperations()
    return function(x) return x + 1 end,
           function(x) return x - 1 end
end

-- Main execution
local function main()
    -- Test regular function
    print(add(5, 3))
    
    -- Test calculator
    Calculator:add(10)
    print(Calculator.value)
    
    -- Test person
    local person = createPerson("Alice")
    print(person:greet())
    
    -- Test counter
    local count = counter()
    print(count())
    print(count())
end

-- Call main
main() 

File: tests/test_parse/data/sample.r
# Regular function
add <- function(a, b) {
  a + b
}

# Function with default arguments
greet <- function(name = "World") {
  paste("Hello,", name, "!")
}

# Function returning multiple values
get_stats <- function(x) {
  list(
    mean = mean(x),
    median = median(x),
    sd = sd(x)
  )
}

# Function with variable arguments
sum_all <- function(...) {
  args <- list(...)
  sum(unlist(args))
}

# Function with formula
regression <- function(formula, data) {
  lm(formula, data = data)
}

# Anonymous function
lapply(1:3, function(x) x^2)

# Closure
make_counter <- function() {
  count <- 0
  function() {
    count <<- count + 1
    count
  }
}

# S3 class and methods
make_person <- function(name, age) {
  structure(
    list(
      name = name,
      age = age
    ),
    class = "person"
  )
}

print.person <- function(x, ...) {
  cat("Person:", x$name, "\n")
  cat("Age:", x$age, "\n")
}

# S4 class and methods
setClass(
  "Calculator",
  slots = list(
    value = "numeric"
  )
)

setMethod(
  "show",
  "Calculator",
  function(object) {
    cat("Calculator value:", object@value, "\n")
  }
)

setGeneric("add_value", function(obj, x) standardGeneric("add_value"))

setMethod(
  "add_value",
  "Calculator",
  function(obj, x) {
    obj@value <- obj@value + x
    obj
  }
)

# Function with attributes
weighted_mean <- function(x, w) {
  sum(x * w) / sum(w)
}
attr(weighted_mean, "description") <- "Calculates weighted mean"

# Main execution
main <- function() {
  # Test regular function
  result <- add(5, 3)
  print(result)
  
  # Test person class
  person <- make_person("Alice", 30)
  print(person)
  
  # Test calculator
  calc <- new("Calculator", value = 0)
  calc <- add_value(calc, 10)
  show(calc)
}

# Run main
main() 